### PR TITLE
Add `sticky-file-header` feature

### LINKED ISSUE #5459: Sticky file header on commits and compare pages
### Description

This is implemented in PRs [way back to 2019](https://github.com/refined-github/refined-github/pull/647#issuecomment-474000547), but other friends of `hasFiles` never got this.

Would you consider it useful? For me a little bit extra context is always helpful.

### Screenshot

_No response_

### Example URL

https://github.com/refined-github/refined-github/commit/1d9dd9f0208c327351512beb6d6a22279bc47807

### PR DESCRIPTION
**Pull request by kidonng**
Resolve #5459

## Test URLs

- https://github.com/refined-github/refined-github/commit/1d9dd9f0208c327351512beb6d6a22279bc47807
- https://github.com//refined-github/refined-github/compare/22.2.13...22.2.22#files_bucket

## Screenshot

https://user-images.githubusercontent.com/44045911/156828581-eea2ce6a-91a0-48da-9634-f64a9b453fd2.mov

### CODE REVIEW

**Review by cheap-glitch — COMMENTED** [review]
If we copy the styles of the four relevant classes from GitHub's stylesheet, this could become pure CSS.

Imo it could even be in `refined-github.css`, since this is improving the overall consistency of the interface, and I don't see anyone wanting to turn this off.

**Comment by kidonng** [comment]
> this could become pure CSS.

Nope, those are `.js-*` classes and they are called that for a reason:

```ts
// https://github.githubassets.com/assets/app/assets/modules/github/behaviors/sticky.ts
observe('.js-sticky', {
  constructor: HTMLElement,
  add(el) {
    initializeSet(el)
  },
  remove(el) {
    removeSet(el)
    manageObservers()
  }
})
```

**Review by fregante — CHANGES_REQUESTED** [review]
> Nope, those are `.js-*` classes and they are called that for a reason:

What's the reason exactly? I don't see it. What I see is that 2 lines of CSS work as expected in this case:

https://user-images.githubusercontent.com/1402241/156872695-60f18f0d-7ade-4d30-9caa-d8d2aa4255fa.mov

**Comment by kidonng** [comment]
> What's the reason exactly? I don't see it. What I see is that 2 lines of CSS work as expected in this case:

> ```ts
> // https://github.githubassets.com/assets/app/assets/modules/github/behaviors/sticky.ts
> // `position: sticky` behavior that works with the notification shelf
> // ...
> // - Ensures that any sticky headers that are on the same page as the notifications shelf sit below it vertically, rather than hidden underneath the shelf
> ```

**Comment by kidonng** [comment]
I forget it needs `selector-observer` for lazy loaded items 🥲

**Comment by cheap-glitch** [comment]
Could we maybe check for the the presence of a notification bar (ideally via CSS) and adjust accordingly?

**Comment by fregante** [comment]
> Could we maybe check for the the presence of a notification bar (ideally via CSS) and adjust accordingly?

Yes:

```css
.file-header {
	position: sticky;
	top: 0;
}
.notification-shelf ~ .application-main .file-header {
	top: 70px;
}
```

Screen Shot 23

gif



@kidonng I hope in the future we can reach these solutions without snark like:

> and they are called that for a reason:

File: source/features/sticky-file-header.css [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	position: sticky;
+	top: 0;
+	/* Same as `.sticky-file-header.is-stuck */
+	z-index: 6;
+}
+
+/* Same as `.sticky-file-header.has-open-dropdown` */
+.file-header.has-open-dropdown:not(.sticky-file-header) {
+	z-index: 10;
+}
```
Comment (cheap-glitch) [inline source/features/sticky-file-header.css:12]: Would it break something if we set it to 10 by default?

File: readme.md [hunk 2]
```diff
@@ -326,6 +326,7 @@ Thanks for contributing! 🦋🙌
 - [](# "easy-toggle-files") [Enables toggling file diffs by clicking on their header bar.](https://user-images.githubusercontent.com/47531779/99855419-be173e00-2b7e-11eb-9a55-0f6251aeb0ef.gif)
 - [](# "same-branch-author-commits") [Preserves current branch and path when viewing all commits by an author.](https://user-images.githubusercontent.com/44045911/148764372-ee443213-e61a-4227-9219-0ee54ed832e8.png)
 - [](# "easy-toggle-commit-messages") [Enables toggling commit messages by clicking on the commit box.](https://user-images.githubusercontent.com/1402241/152121837-ca13bf8a-9b7f-4517-8e8d-b58bb135523b.gif)
+- [](# "sticky-file-header") [Enables sticky file header on commits and compare pages.](https://user-images.githubusercontent.com/44045911/156827766-283e8388-7053-418f-a38b-16f615c6af4a.gif)
```
Comment (cheap-glitch) [inline readme.md:329]:
```suggestion
- [](# "sticky-file-header") [Enables sticky file headers on commit and compare pages.](https://user-images.githubusercontent.com/44045911/156827766-283e8388-7053-418f-a38b-16f615c6af4a.gif)
```

**Review by cheap-glitch — APPROVED** [review]
Successfully tested on Firefox with & without the notification bar.

**Review by fregante — APPROVED** [review]
💙

File: source/features/sticky-file-header.css (hunk 1 above)
Comment (kidonng) [reply source/features/sticky-file-header.css:12]: I don't want to mess with z-index, make it always 6 is already a bit 😬

File: readme.md (hunk 2 above)
Comment (kidonng) [reply readme.md]: 🥲 actually this should not be in the README anymore

**Pull request merged**